### ISSUE TITLE
Can't dnd in outline palette to the end of the tree

### ISSUE THREAD

**Issue by doronrosenberg** [maqetta/maqetta#2837]
Happens in m6 as well

**Comment by doronrosenberg** [comment]
@JonFerraiolo 

According to blame you added this to reparentcommand

```
// If moving a widget within same parent, adjust newIndex in case the widget is being moved
    43  // to a latter point in list of children. If so, the removeChild operation has altered the child list
    44  var newIndex = (newParent == oldParent && this._oldIndex < this._newIndex) ? this._newIndex -1 : this._newIndex;
    45  newParent.addChild(widget, newIndex);
```

That -1 is breaking the dnd.  Removing it fixes the issue and I don't see why that even exists.  Do you perhaps remember why?

**Comment by JonFerraiolo** [comment]
OK, I studied things, and here's the deal. 

If a widget is being reparented where the old parent and new parent are one in the same, and the widget is being moved to a newindex which is greater than oldindex, then the removal of the widget from its old position will shift all subsequent children one position forward, which means newIndex needs to be deprecated by one to take into account all of this shifting. 

The problem with the Outline dnd scenario is that there are two things that are doing the deprecate-by-one adjustment: dndSource.js and ReparentCommand.js. Here is the code from dndSource.js that decreases index by 1:

```
                if(source == this){
                    // This is a node from my own tree, and we are moving it, not copying.
                    // Remove item from old parent's children attribute.
                    // TODO: dijit.tree.dndSelector should implement deleteSelectedNodes()
                    // and this code should go there.

                    if(typeof insertIndex == "number"){
                        if(newParentItem == oldParentItem && childTreeNode.getIndexInParent() < insertIndex){
                            insertIndex -= 1;
                        }
                    }
                    model.pasteItem(childItem, oldParentItem, newParentItem, copy, insertIndex, before);
```

Clearly, we need to specify unambiguously what ReparentCommand does in the case of same parent and oldIndex < newIndex. Should the calling routine do the deprecate-by-one logic, or should ReparentCommand do the deprecate-by-one?

ReparentCommand.js is called from multiple places - outline tree dnd is just one of many. In fact, a quick `grep ... | wc -l` shows 19 places that use ReparentCommand. If we remove the deprecate-by-1 logic from ReparentCommand for this "same parent" case, then we will have to include the deprecate-by-1 logic to a higher level in some of those 19 scenarios. Note that I added the deprecrate-by-one logic when implementing the various sketch-mode arrange commands (move-to-front, move-to-back), and we've been doing a lot of testing in the past few months with the visual page editor drag/drop. I suspect most of these 19 scenarios are coded to expect ReparentCommand.js to do the deprecate-by-one logic.

My recommendation is:
- include logic in the `pasteItem` function within VisualEditorOutline.js to undo the deprecate-by-one adjustment that dndSource has done, with an appropriate comment explaining why this is necessary
- add comments to ReparentCommand.js that explicitly says that in the case of same parent and oldIndex<newIndex, the newIndex value should be the target position within the original child list, not the position after the widgets have been rearranged.
